weave/trace/op.py

```diff
@@ -256,3 +62,264 @@ def _apply_fn_defaults_to_inputs(
             elif param.kind == inspect.Parameter.VAR_KEYWORD:
                 inputs[param_name] = dict()
     return inputs
+
+
+@runtime_checkable
+class Op(Protocol):
+    """
+    The interface for Op-ified functions and methods.
+
+    Op was previously a class, and has been converted to a Protocol to allow
+    functions to pass for Op.  This is needed because many popular packages are
+    using the `inspect` module for control flow, and Op instances don't always
+    pass those checks.  In particular, `inspect.iscoroutinefunction` always
+    fails for classes, even ones that implement async methods or protocols.
+
+    Some of the attributes are carry-overs from when Op was a class.  We should
+    consider removing the unnecessary ones where possible.
+    """
+
+    name: str
+    signature: inspect.Signature
+    ref: Optional[ObjectRef]
+    resolve_fn: Callable
+
+    call: Callable[..., Any]
+    calls: Callable[..., "CallsIter"]
+
+    # not sure if this is the best place for this, but kept for compat
+    _set_on_output_handler: Callable[[OnOutputHandlerType], None]
+    _on_output_handler: Optional[OnOutputHandlerType]
+
+    __call__: Callable[..., Any]
+    __self__: Any
+
+
+def _set_on_output_handler(func: Op, on_output: OnOutputHandlerType) -> None:
+    if func._on_output_handler is not None:
+        raise ValueError("Cannot set on_output_handler multiple times")
+    func._on_output_handler = on_output
+
+
+def _is_method_alt(func: Callable) -> bool:
```

Comment:
what does this function do?

Reply:
clarified as `_is_unbound_method` in `op.py`